### PR TITLE
fix: invalid yield generated

### PR DESCRIPTION
**Pull request by maikelordaz**

### CODE REVIEW

File: contracts/libraries/LibYieldGeneration.sol [hunk 1]
```diff
@@ -93,14 +93,13 @@ library LibYieldGeneration {
...
+    function _neededShares(
         uint _collateralAmount,
         LibYieldGenerationStorage.YieldGeneration storage _yield
     ) internal view returns (uint) {
-        uint pricePerShare = IZaynVaultV2TakaDao(_yield.providerAddresses["ZaynVault"])
-            .getPricePerFullShare();
-
-        return ((_collateralAmount * 10 ** 18) / pricePerShare);
+        return
+            (((_collateralAmount * 10 ** 18) / _yield.totalDeposit) * _yield.totalShares) /
```
Comment (MohammedTaka) [inline contracts/libraries/LibYieldGeneration.sol]: Could we get rid of the decimal multiplication and divisions by doing:

`_collateralAmount * _yield.totalShares /  _yield.totalDeposit`

File: contracts/libraries/LibYieldGeneration.sol [hunk 2]
```diff
@@ -93,14 +93,13 @@ library LibYieldGeneration {
     /// @notice Conversion from eth to shares
     /// @param _collateralAmount The amount of collateral to withdraw
     /// @param _yield The yield generation struct
-    function _ethToShares(
+    function _neededShares(
         uint _collateralAmount,
         LibYieldGenerationStorage.YieldGeneration storage _yield
```
Comment (MohammedTaka) [inline contracts/libraries/LibYieldGeneration.sol]: Can this not be memory instead of storage? We are not writing to _yield here.

**Review by MohammedTaka — CHANGES_REQUESTED** [review]
Logic is solid, although some optimizations can be made. I'm also missing unit tests that test the withdrawal of yield

File: contracts/libraries/LibYieldGeneration.sol (hunk 1 above)
Comment (maikelordaz) [reply contracts/libraries/LibYieldGeneration.sol]: Commit 6044c03

File: contracts/libraries/LibYieldGeneration.sol (hunk 2 above)
Comment (maikelordaz) [reply contracts/libraries/LibYieldGeneration.sol]: No it can't because of the nested mapping, But changed only to the needed inputs on commit 89373d2

**Comment by MohammedTaka** [comment]
@maikelordaz could you simulate the the reimbursement code with the inputs from the google doc on jira?

File: contracts/facets/YGFacetZaynFi.sol [hunk 3]
```diff
@@ -119,4 +122,110 @@ contract YGFacetZaynFi is IYGFacetZaynFi {
...
+                    vaultAddress,
+                    termId
+                );
+
+                // Increment the used value so far
+                usedValue += neededEth;
+
+                // Validate the amount of shares deposited
+                uint sharesAfter = IZaynVaultV2TakaDao(vaultAddress).balanceOf(termId);
+                require(neededShares == (sharesAfter - sharesBefore), "Invalid amount of shares deposited");
```
Comment (maikelordaz) [inline contracts/facets/YGFacetZaynFi.sol]: There is a little difference that fails here. For example for the first user in the inputs the neededShares are 158058370216700 and the sharesAfter - sharesBefore are 157458744451636

File: contracts/facets/YGFacetZaynFi.sol [hunk 4]
```diff
@@ -119,4 +122,110 @@ contract YGFacetZaynFi is IYGFacetZaynFi {
...
+                
+            } else if (reimbursement < 0) {
+                // When there is a negative reimbursement, we compensate the pool by adding back the exact amount of shares that were lost
+                uint neededShares = uint(reimbursement * -1);
+
+                // Calculate the amount of eth we need to deposit to get the desired shares
+                uint pricePerShare = IZaynVaultV2TakaDao(vaultAddress)
+                    .getPricePerFullShare();
+
+                uint neededEth = neededShares * pricePerShare;
```
Comment (maikelordaz) [inline contracts/facets/YGFacetZaynFi.sol]: Here I fixed the decimals with neededShares * pricePerShare / 10 ** 18

File: contracts/facets/YGFacetZaynFi.sol (hunk 3 above)
Comment (MohammedTaka) [reply contracts/facets/YGFacetZaynFi.sol]: What we can do is we should ZapIn 2 times as much eth, then zapput the difference in before and after so we have an exact amount of shares zapped in

File: contracts/facets/YGFacetZaynFi.sol (hunk 4 above)
Comment (MohammedTaka) [reply contracts/facets/YGFacetZaynFi.sol]: Yeah forgot that, my bad. Please commit

File: contracts/facets/YGFacetZaynFi.sol [hunk 5]
```diff
@@ -211,16 +218,16 @@ contract YGFacetZaynFi is IYGFacetZaynFi {
...
                 );
 
-                (bool successWithdrawExtraShares, ) = payable(msg.sender).call{
-                    value: withdrawExtraShares
-                }("");
-                require(successWithdrawExtraShares, "Failed to send extra shares back");
+                // Give the extra eth back to msg.sender
+                usedValue -= withdrawnExtraEth;
+
+                emit OnYieldCompensated(termId, user, neededEth);
```
Comment (MohammedTaka) [inline contracts/facets/YGFacetZaynFi.sol]: The third parameter should be: neededEth - withdrawnExtraEth

File: contracts/facets/YGFacetZaynFi.sol [hunk 6]
```diff
@@ -235,6 +242,118 @@ contract YGFacetZaynFi is IYGFacetZaynFi {
...
+
+                // Give the extra eth back to msg.sender
+                usedValue -= withdrawnExtraEth;
+            }
+
+            unchecked {
+                ++i;
+            }
+
+            emit OnYieldTermUpdated(termId);
```
Comment (MohammedTaka) [inline contracts/facets/YGFacetZaynFi.sol]: This one seems redundant, maybe add how much ETH was restored and how much was compensated. So add: withdrawnTooMuch, (neededEth - withdrawnExtraEth) & the final balanceOf from zaynfi

**Review by MohammedTaka — CHANGES_REQUESTED** [review]
Looking good, some minor points and I think we're good to go

**Comment by maikelordaz** [comment]
@MohammedTaka I add new checks when withdrawing collateral on the unit tests

**Comment by MohammedTaka** [comment]
Added an additional function to make compensate for the additional yield that was reimbursed with a zaynfi bug.

**Review by MohammedTaka — APPROVED** [review]

**Review by MohammedTaka — COMMENTED** [review]
We need to make sure withdrawing shares does not cause rounding errors

**Review by elallamTaka — APPROVED** [review]
BismilAllah

**Pull request merged**